### PR TITLE
Attach spans to generated Hydroflow code in Dedalus

### PR DESCRIPTION
**Pull request by shadaj**
Attach spans to generated Hydroflow code in Dedalus

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/hydro-project/hydroflow/pull/518).
* #522
* __->__ #518

### CODE REVIEW

File: hydroflow_datalog_core/src/lib.rs [hunk 1]
```diff
@@ -24,16 +27,21 @@ pub fn parse_pipeline(
...
+
+    syn::LitStr::new(&code_str.value, subspan)
+        .parse()
+        .map_err(|err| {
+            vec![Diagnostic {
+                span: subspan,
+                level: Level::Error,
+                message: format!("Failed to parse input pipeline: {}", err),
+            }]
+        })
```
Comment (MingweiSamuel) [inline hydroflow_datalog_core/src/lib.rs:44]: This ought to use `get_span` too? Guess there wouldn't be a way to incorporate `err.span()` though then

Also need to move that commend down to line 39

**Review by MingweiSamuel — APPROVED** [review]
I guess the spanned idents will cause errors if you ever try to call `datalog!` from a declarative macro fwiw

**Comment by shadaj** [comment]
@MingweiSamuel could use a second pass, made spans more precise and adjusted the order of subgraph emission in Hydroflow to improve type inference (rather than complaining about inputs being the wrong type, we now shift the blame to joins operating on incompatible types).

**Review by MingweiSamuel — APPROVED** [review]

**Comment by MingweiSamuel** [comment]
It would also be good to add compile-fail tests for datalog

File: hydroflow_datalog_core/src/join_plan.rs [hunk 2]
```diff
@@ -128,11 +131,11 @@ fn emit_join_input_pipeline(
 /// For example, `rel(a, b, a) := ...` requires that the values in the 0th and 2nd slots be the
 /// same, so we would return a map `{ "a" => [0, 2] }`. Note that since `b` is not repeated, it is
 /// not in the map.
-fn find_relation_local_constraints(
-    fields: &[crate::grammar::datalog::Ident],
+fn find_relation_local_constraints<'a>(
+    fields: impl Iterator<Item = &'a Spanned<crate::grammar::datalog::Ident>>,
```
[the remark this answers is not in the document]
Comment (shadaj) [reply hydroflow_datalog_core/src/join_plan.rs:135]: Anonymous lifetimes in impl traits are unstable 😢

**Pull request merged**